src/lib/expression/evaluation/expression_evaluator.cpp

```diff
@@ -0,0 +1,1054 @@
+  });
+}
+
+template <typename Functor>
+void ExpressionEvaluator::_resolve_to_expression_result(const AbstractExpression& expression, const Functor& fn) {
+  Assert(expression.type != ExpressionType::List, "Can't resolve ListExpression to ExpressionResult");
+
+  if (expression.data_type() == DataType::Null) {
+    // resolve_data_type() doesn't support Null, so we have handle it explicitly
+    ExpressionResult<NullValue> null_value_result{{NullValue{}}, {true}};
+
+    fn(null_value_result);
+
+  } else {
+    resolve_data_type(expression.data_type(), [&](const auto data_type_t) {
+      using ExpressionDataType = typename decltype(data_type_t)::type;
+
+      const auto expression_result = evaluate_expression_to_result<ExpressionDataType>(expression);
+      fn(*expression_result);
+    });
+  }
+}
+
+template <typename... RowCounts>
+ChunkOffset ExpressionEvaluator::_result_size(const RowCounts... row_counts) {
+  // If any operand is empty (that's the case IFF it is an empty column) the result of the expression has no rows
+  if (((row_counts == 0) || ...)) return 0;
+
+  return std::max({row_counts...});
+}
+
+std::vector<bool> ExpressionEvaluator::_evaluate_default_null_logic(const std::vector<bool>& left,
+                                                                    const std::vector<bool>& right) const {
+  const auto result_size = _result_size(left.size(), right.size());
+
+  if (result_size == 0) return {};
+
+  if (left.size() == right.size()) {
+    std::vector<bool> nulls(result_size);
+    std::transform(left.begin(), left.end(), right.begin(), nulls.begin(), [](auto l, auto r) { return l || r; });
+    return nulls;
+  } else if (left.size() > right.size()) {
+    DebugAssert(right.size() == 1,
+                "Operand should have either the same row count as the other or 1 (to represent a literal)");
+    if (right.front())
+      return std::vector<bool>({true});
+    else
+      return left;
+  } else {
+    DebugAssert(left.size() == 1,
+                "Operand should have either the same row count as the other or 1 (to represent a literal)");
+    if (left.front())
+      return std::vector<bool>({true});
+    else
+      return right;
+  }
+}
+
+void ExpressionEvaluator::_ensure_column_materialization(const ColumnID column_id) {
+  Assert(_chunk, "Cannot access Columns in this Expression as it doesn't operate on a Table/Chunk");
+
+  if (_column_materializations[column_id]) return;
+
+  const auto& column = *_chunk->get_column(column_id);
+
+  resolve_data_type(column.data_type(), [&](const auto column_data_type_t) {
+    using ColumnDataType = typename decltype(column_data_type_t)::type;
+
+    std::vector<ColumnDataType> values;
+    materialize_values(column, values);
+
+    if (_table->column_is_nullable(column_id)) {
+      std::vector<bool> nulls;
+      materialize_nulls<ColumnDataType>(column, nulls);
```

Comment:
I'm not a fan of these out-parameters. Why wouldn't `auto = materialize_nulls<...>(column)` work here?

Reply:
Not part of this PR imo, and I don't like the design of `materialize_...()` that much anymore either. The reason we made it an out parameter back then was that we could implement materizalize_nulls() somewhat container type independent. 
Will gladly polish our materialization code, but would do so in a separate issue, if you open one.